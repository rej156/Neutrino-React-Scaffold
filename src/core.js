import clone from 'ramda/src/clone';
import uniq from 'ramda/src/uniq';
import isNil from 'ramda/src/isNil';

// Core API
const core = {
  clear: {},
  get: {},
  register: {},
};
export default core;

// Top-level application items
// the initial value determines how items are registered / updated
// - for []: allows any item & register "uniquely" concatenates
// - for {}: only allows object items & register merges
// - for null | undefined: allows any item & register replacea
const coreItems = {
  componentNeeds: {},
  initialState: {},
  persistedStateKeys: [],
  preRenderActions: [],
  reducers: {},
  rootComponent: null,
  router: null,
  routesHandlers: [],
  store: null,
  storeMiddleware: [],
};

const isObject = val =>
  val !== null && !Array.isArray(val) && val instanceof Object && typeof val === 'object';

class CoreItem {
  constructor(key) {
    this.key = key;
    this.initialValue = this.get();
  }

  register = (value, shouldReplace) => {
    const key = this.key;
    const currentValue = this.get();

    const done = updatedValue => {
      coreItems[key] = updatedValue;

      return core.register;
    };

    if (isNil(value)) {
      return done(shouldReplace ? this.initialValue : currentValue);
    }

    if (Array.isArray(this.initialValue)) {
      return done(shouldReplace ? [].concat(value) : uniq(currentValue.concat(value)));
    }

    if (isObject(this.initialValue)) {
      if (!isObject(value) && !shouldReplace) {
        throw new Error(
          `core.register.${key}() expects an object, you passed ${JSON.stringify(value)}.`,
        );
      }

      return done(shouldReplace ? value : { ...currentValue, ...value });
    }

    return done(value);
  };

  clear = () => {
    this.register(null, true);

    return core.clear;
  };

  get = () => clone(coreItems[this.key]);
}

/**
 * All core item keys
 */
export const itemKeys = Object.keys(coreItems);

/**
 * Assign getter & setter functions for core items, e.g:
 * - core.register.store(createStore(...))
 * - core.clear.store()
 * - core.get.store()
 */
itemKeys.forEach(key => {
  const item = new CoreItem(key);

  core.register[key] = item.register;
  core.clear[key] = item.clear;
  core.get[key] = item.get;
});
