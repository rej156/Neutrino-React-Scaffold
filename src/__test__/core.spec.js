import clone from 'ramda/src/clone'
import core, { itemKeys } from '../core'

describe('@monty/core', () => {
  const mockObjectValues = { foo: '', bar: {}, baz: [], qux: () => {} }
  const mockArrayValues = ['foo', {}, [], () => {}]

  afterEach(() => {
    core.clear
      .store()
      .componentNeeds()
      .routesHandlers()
  })

  it('has registered items', () => {
    expect(itemKeys).toMatchSnapshot()
  })

  describe('.register', () => {
    it('exposes a method for each core item', () => {
      expect(core.register).toMatchSnapshot()
    })

    describe('method for items initialised with array content', () => {
      it('concatenates the content value to the item array', () => {
        core.register.routesHandlers({ foo: () => {} })
        const before = core.get.routesHandlers()
        core.register.routesHandlers({ bar: () => {} })
        const after = core.get.routesHandlers()

        expect({ before, after }).toMatchSnapshot()
      })
      it('replaces the item with the content array if shouldReplace', () => {
        core.register.routesHandlers([{ baz: () => {} }, { qux: () => {} }])
        const before = core.get.routesHandlers()
        core.register.routesHandlers({ foo: () => {} }, true)
        const after = core.get.routesHandlers()

        expect({ before, after }).toMatchSnapshot()
      })
      it('replaces the item with an empty array if shouldReplace and content is nil', () => {
        core.register.routesHandlers({ baz: () => {} })
        const before = core.get.routesHandlers()
        core.register.routesHandlers(undefined, true)
        const after = core.get.routesHandlers()

        expect({ before, after }).toMatchSnapshot()
      })
    })

    describe('method for items initialised with plain object content', () => {
      it('throws if not called with a plain object', () => {
        expect(() => {
          core.register.componentNeeds(['foo'])
        }).toThrow(/register\.componentNeeds.*\["foo"\]/)
      })

      it('merges the content object with the item object', () => {
        core.register.componentNeeds({ foo: () => {}, bar: () => {} })
        const before = core.get.componentNeeds()
        core.register.componentNeeds({ baz: () => {} })
        const after = core.get.componentNeeds()

        expect({ before, after }).toMatchSnapshot()
      })

      it('replaces the item with the content object if shouldReplace', () => {
        core.register.componentNeeds({ baz: () => {}, qux: () => {} })
        const before = core.get.componentNeeds()
        core.register.componentNeeds({ foo: () => {} }, true)
        const after = core.get.componentNeeds()

        expect({ before, after }).toMatchSnapshot()
      })

      it('replaces the item with an empty object if shouldReplace and content is nil', () => {
        core.register.componentNeeds({ baz: () => {} })
        const before = core.get.componentNeeds()
        core.register.componentNeeds(undefined, true)
        const after = core.get.componentNeeds()

        expect({ before, after }).toMatchSnapshot()
      })
    })

    describe('method for items initialised with nil (any content)', () => {
      it('always replaces the item with the content value', () => {
        core.register.store({ baz: () => {} })
        const store1 = core.get.store()
        core.register.store(() => {})
        const store2 = core.get.store()
        core.register.store({})
        const store3 = core.get.store()

        expect({ store1, store2, store3 }).toMatchSnapshot()
      })
    })
  })

  describe('.clear', () => {
    it('exposes a method for each core item', () => {
      expect(core.clear).toMatchSnapshot()
    })

    describe('method for items with array content', () => {
      it('replaces the item with initial value', () => {
        const initialValue = core.get.routesHandlers()
        core.register.routesHandlers(mockArrayValues)
        core.clear.routesHandlers()
        expect(core.get.routesHandlers()).toEqual(initialValue)
      })
    })
    describe('method for items with plain object content', () => {
      it('replaces the item with initial value', () => {
        const initialValue = core.get.componentNeeds()
        core.register.componentNeeds(mockObjectValues)
        core.clear.componentNeeds()
        expect(core.get.componentNeeds()).toEqual(initialValue)
      })
    })
    describe('method for items with any content', () => {
      it('replaces the item with initial value', () => {
        const initialValue = core.get.store()
        core.register.store({})
        core.clear.store()
        expect(core.get.store()).toEqual(initialValue)
      })
    })
  })

  describe('.get', () => {
    it('exposes a method for each core item', () => {
      expect(core.get).toMatchSnapshot()
    })

    describe('method for items with array content', () => {
      it('returns the cloned value', () => {
        core.register.routesHandlers(clone(mockArrayValues))
        core.get.routesHandlers()[0] = 'foobarbazqux'
        expect(core.get.routesHandlers()).toEqual(mockArrayValues)
      })
    })
    describe('method for items with plain object content', () => {
      it('returns the cloned value', () => {
        core.register.componentNeeds(clone(mockObjectValues))
        core.get.componentNeeds().foobarbazqux = true
        expect(core.get.componentNeeds()).toEqual(mockObjectValues)
      })
    })
    describe('method for items with any content', () => {
      it('returns a cloned value', () => {
        const mockFunc = () => {}
        core.register.store(mockFunc)
        expect(core.get.store()).toEqual(mockFunc)

        core.register.store(clone(mockArrayValues))
        core.get.store()[0] = 'foobarbazqux'
        expect(core.get.store()).toEqual(mockArrayValues)

        core.register.store(clone(mockObjectValues))
        core.get.store().foobarbazqux = true
        expect(core.get.store()).toEqual(mockObjectValues)
      })
    })
  })
})
