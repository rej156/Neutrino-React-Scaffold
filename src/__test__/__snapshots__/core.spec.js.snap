// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@monty/core .clear exposes a method for each core item 1`] = `
Object {
  "componentNeeds": [Function],
  "initialState": [Function],
  "persistedStateKeys": [Function],
  "preRenderActions": [Function],
  "reducers": [Function],
  "rootComponent": [Function],
  "router": [Function],
  "routesHandlers": [Function],
  "sagas": [Function],
  "store": [Function],
  "storeMiddleware": [Function],
}
`;

exports[`@monty/core .get exposes a method for each core item 1`] = `
Object {
  "componentNeeds": [Function],
  "initialState": [Function],
  "persistedStateKeys": [Function],
  "preRenderActions": [Function],
  "reducers": [Function],
  "rootComponent": [Function],
  "router": [Function],
  "routesHandlers": [Function],
  "sagas": [Function],
  "store": [Function],
  "storeMiddleware": [Function],
}
`;

exports[`@monty/core .register exposes a method for each core item 1`] = `
Object {
  "componentNeeds": [Function],
  "initialState": [Function],
  "persistedStateKeys": [Function],
  "preRenderActions": [Function],
  "reducers": [Function],
  "rootComponent": [Function],
  "router": [Function],
  "routesHandlers": [Function],
  "sagas": [Function],
  "store": [Function],
  "storeMiddleware": [Function],
}
`;

exports[`@monty/core .register method for items initialised with array content concatenates the content value to the item array 1`] = `
Object {
  "after": Array [
    Object {
      "foo": [Function],
    },
    Object {
      "bar": [Function],
    },
  ],
  "before": Array [
    Object {
      "foo": [Function],
    },
  ],
}
`;

exports[`@monty/core .register method for items initialised with array content replaces the item with an empty array if shouldReplace and content is nil 1`] = `
Object {
  "after": Array [],
  "before": Array [
    Object {
      "baz": [Function],
    },
  ],
}
`;

exports[`@monty/core .register method for items initialised with array content replaces the item with the content array if shouldReplace 1`] = `
Object {
  "after": Array [
    Object {
      "foo": [Function],
    },
  ],
  "before": Array [
    Object {
      "baz": [Function],
    },
    Object {
      "qux": [Function],
    },
  ],
}
`;

exports[`@monty/core .register method for items initialised with nil (any content) always replaces the item with the content value 1`] = `
Object {
  "store1": Object {
    "baz": [Function],
  },
  "store2": [Function],
  "store3": Object {},
}
`;

exports[`@monty/core .register method for items initialised with plain object content merges the content object with the item object 1`] = `
Object {
  "after": Object {
    "bar": [Function],
    "baz": [Function],
    "foo": [Function],
  },
  "before": Object {
    "bar": [Function],
    "foo": [Function],
  },
}
`;

exports[`@monty/core .register method for items initialised with plain object content replaces the item with an empty object if shouldReplace and content is nil 1`] = `
Object {
  "after": Object {},
  "before": Object {
    "baz": [Function],
  },
}
`;

exports[`@monty/core .register method for items initialised with plain object content replaces the item with the content object if shouldReplace 1`] = `
Object {
  "after": Object {
    "foo": [Function],
  },
  "before": Object {
    "baz": [Function],
    "qux": [Function],
  },
}
`;

exports[`@monty/core has registered items 1`] = `
Array [
  "store",
  "router",
  "rootComponent",
  "preRenderActions",
  "componentNeeds",
  "routesHandlers",
  "storeMiddleware",
  "persistedStateKeys",
  "initialState",
  "reducers",
  "sagas",
]
`;
